### PR TITLE
Add a documentation chapter summarizing plugin points

### PR DESCRIPTION
**Pull request by benclifford**
## Type of change

- Documentation update

### CODE REVIEW

File: docs/userguide/plugins.rst [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+several implementations that use each plugin point.
+
+This page gives a brief summary of those places and why you might want
+to use them, with links to the API guide.
+
+Executors
+---------
+When the parsl data flow kernel is ready for a task to run, it passes that
+task to an `Executor`. The executor is then responsible for running the task's
+python code and returning the result. This is the abstraction that allows one
```
Comment (danielskatz) [inline docs/userguide/plugins.rst:14]:
```suggestion
Python code and returning the result. This is the abstraction that allows one
```

File: docs/userguide/plugins.rst [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+execute tasks.
+
+The particular way in which a system makes those workers start is implemented
+by `Providers` and `Launchers`.
+
+A `Provider` allows a command line to be submitted as a request to the
+underlying batch system to be run inside an allocation of nodes.
+
+A `Launcher` modifies that command line when run inside the allocation to
+add on any wrappers that are needed to launch the command. (eg srun inside
```
Comment (danielskatz) [inline docs/userguide/plugins.rst:33]:
```suggestion
add on any wrappers that are needed to launch the command (eg srun inside
```

File: docs/userguide/plugins.rst [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+
+Memoization/checkpointing
+-------------------------
+
+When parsl memoizes/checkpoints an app parameter, it does so by computing a
+hash of that parameter that should be the same if that parameter is the same
+on subsequent invocations. This isn't straightforward to do for arbitrary
+objects, so parsl implements a checkpointing hash function for a few common
+types, and raises an exception on unknown types (TK put in unknown exception
```
Comment (danielskatz) [inline docs/userguide/plugins.rst:59]: I'm not sure what TK means here

Comment (benclifford) [reply docs/userguide/plugins.rst:59]: https://en.wikipedia.org/wiki/To_come_(publishing)

**Review by yadudoc — APPROVED** [review]
@benclifford This looks good. There's a note in the docs to add an example, are you planning to add that separately?

**Comment by benclifford** [comment]
@yadudoc I put in the example that I missed before

**Review by yadudoc — APPROVED** [review]
Let's merge this!

**Comment by yadudoc** [comment]
@benclifford Here's the error in testing:
```
Warning, treated as error:
/home/runner/work/parsl/parsl/docs/userguide/plugins.rst:61:Error in "code-block" directive:
maximum 1 argument(s) allowed, 5 supplied.

.. code-block::
  ValueError("unknown type for memoization ...")
Makefile:52: recipe for target 'html' failed
make: *** [html] Error 2
Error: Process completed with exit code 2.
```

**Pull request merged**